File: src/main/java/com/leijendary/spring/boot/template/core/error/DataIntegrityViolationExceptionHandler.java
package com.leijendary.spring.boot.template.core.error;

import static com.leijendary.spring.boot.template.core.util.RequestContext.getLocale;
import static com.leijendary.spring.boot.template.core.util.StringUtil.snakeCaseToCamelCase;
import static org.apache.commons.lang3.StringUtils.substringBetween;
import static org.springframework.http.HttpStatus.CONFLICT;
import static org.springframework.http.HttpStatus.INTERNAL_SERVER_ERROR;

import com.leijendary.spring.boot.template.core.data.ErrorResponse;

import org.hibernate.exception.ConstraintViolationException;
import org.springframework.context.MessageSource;
import org.springframework.core.annotation.Order;
import org.springframework.dao.DataIntegrityViolationException;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.RestControllerAdvice;

import lombok.RequiredArgsConstructor;

@RestControllerAdvice
@Order(3)
@RequiredArgsConstructor
public class DataIntegrityViolationExceptionHandler {

    private final MessageSource messageSource;

    @ExceptionHandler(DataIntegrityViolationException.class)
    public ResponseEntity<ErrorResponse> catchDataIntegrityViolation(final DataIntegrityViolationException exception) {
        final var errorResponse = getResponse(exception);
        final var status = errorResponse.getMeta().get("status");

        return ResponseEntity
                .status((int) status)
                .body(errorResponse);
    }

    private ErrorResponse getResponse(final DataIntegrityViolationException exception) {
        final var cause = exception.getCause();

        if (cause instanceof ConstraintViolationException) {
            return constraintViolationException((ConstraintViolationException) cause);
        }

        final var code = "error.dataIntegrity";
        final var arguments = new Object[] { exception.getMessage() };
        final var message = messageSource.getMessage(code, arguments, getLocale());

        return ErrorResponse.builder()
                .addError("Entity", code, message)
                .status(INTERNAL_SERVER_ERROR)
                .build();
    }

    private ErrorResponse constraintViolationException(final ConstraintViolationException exception) {
        final var errorMessage = exception.getSQLException().getNextException().getMessage();
        final var field = snakeCaseToCamelCase(substringBetween(errorMessage, "Key (", ")="));
        final var value = substringBetween(errorMessage, "=(", ") ");
        final var code = "validation.alreadyExists";
        final var arguments = new Object[] { field, value };
        final var message = messageSource.getMessage(code, arguments, getLocale());

        return ErrorResponse.builder()
                .addError(field, code, message)
                .status(CONFLICT)
                .build();
    }
}

File: src/main/kotlin/com/leijendary/spring/boot/template/core/error/DataIntegrityViolationExceptionHandler.kt
package com.leijendary.spring.boot.template.core.error

import com.leijendary.spring.boot.template.core.data.ErrorData
import com.leijendary.spring.boot.template.core.extension.snakeCaseToCamelCase
import com.leijendary.spring.boot.template.core.util.RequestContext.locale
import org.apache.commons.lang3.StringUtils.substringBetween
import org.hibernate.exception.ConstraintViolationException
import org.springframework.context.MessageSource
import org.springframework.core.annotation.Order
import org.springframework.dao.DataIntegrityViolationException
import org.springframework.http.HttpStatus.CONFLICT
import org.springframework.http.HttpStatus.INTERNAL_SERVER_ERROR
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.RestControllerAdvice

@RestControllerAdvice
@Order(3)
class DataIntegrityViolationExceptionHandler(private val messageSource: MessageSource) {
    @ExceptionHandler(DataIntegrityViolationException::class)
    fun catchDataIntegrityViolation(exception: DataIntegrityViolationException): ResponseEntity<List<ErrorData>> {
        return getResponse(exception)
    }

    private fun getResponse(exception: DataIntegrityViolationException): ResponseEntity<List<ErrorData>> {
        val cause: Throwable? = exception.cause

        if (cause is ConstraintViolationException) {
            return constraintViolationException(cause)
        }

        val code = "error.dataIntegrity"
        val arguments: Array<String> = arrayOf(exception.message ?: "")
        val message = messageSource.getMessage(code, arguments, locale)
        val errorData = ErrorData(listOf("data", "entity"), code, message)

        return ResponseEntity
            .status(INTERNAL_SERVER_ERROR)
            .body(listOf(errorData))
    }

    private fun constraintViolationException(exception: ConstraintViolationException): ResponseEntity<List<ErrorData>> {
        val errorMessage = exception.sqlException.nextException.message
        val sql: String = exception.sql
        val table: String = sql.let {
            substringBetween(it, "insert into ", " (") ?: substringBetween(it, "update ", " set ")
        }.snakeCaseToCamelCase(true)
        val field: String = substringBetween(errorMessage, "Key (", ")=").snakeCaseToCamelCase()
        val value = substringBetween(errorMessage, "=(", ") ")
        val code = "validation.alreadyExists"
        val arguments = arrayOf(field, value)
        val message = messageSource.getMessage(code, arguments, locale)
        val errorData = ErrorData(listOf("data", table, field), code, message)

        return ResponseEntity
            .status(CONFLICT)
            .body(listOf(errorData))
    }
}